docs/rate-limit-and-backpressure.md

```diff
@@ -0,0 +1,122 @@
+```text
+Backfill (A) -> Stream Materialize (B)
+```
+
+If B is busy, it just won't request new Stream Messages (Barriers and StreamChunks) from A.
+This is called Backpressure. You can see that A's stream will be backpressured by B, i.e. no longer progress, until B is ready again.
+
+Now consider if B had the following sequence of events:
+```text
+1. Scheduled Barrier X (<1ms processing time)
+2. StreamChunk (250ms processing time)
+3. Scheduled Barrier Y (<1ms processing time)
+```
+
+The barrier latency for Y will be 0ms, since our barrier interval is 256ms,
+and the time between X and Y is less than 256ms.
+
+Consider another case where B has the following sequence of events:
+```text
+1. Scheduled Barrier X (<1ms processing time)
+2. StreamChunk (256ms processing time)
+3. StreamChunk (250ms processing time)
+4. Scheduled Barrier Y (<1ms processing time)
+```
+
+Now barrier latency will be around 250ms.
+
+In the case above, where we have Backfill and Stream Materialize, this is unlikely to occur however.
+Backfill should always let upstream side (barriers) bypass, so we should see stream chunk (3) only after (4).
+
+### Adding Channels to the equation
+
+Now consider a case where we have `join`.
+
+```
+StreamScan A -- HashJoin X
+              /
+             /
+StreamScan B
+```
+
+Now both A and B will have a channel between themselves and X.
+The channel is there to provide a buffer between actors, since A and B are not directly connected to X, they could be on different machines.
+Hence providing a buffer isolates X from network latency of getting Stream Messages from A and B.
+
+If X is busy, it will first backpressure the channel.
+However, the channel will not immediately backpressure A and B.
+It will only do so once it is full.
+
+This means that the stream chunks between barriers are equivalent to channel capacity.
+And then the barrier latency is the duration taken to process these barriers.
+
+_example.toml_
+```toml
+[streaming.developer]
+stream_exchange_initial_permits = 2048
+```
+
+### Considering Amplification and UDF latency
+
+When there's high amplification or UDF latency in some downstream actor, upstream will naturally be backpressured.
+Overtime, a stream graph will hit its max capacity (all actor channels are full).
+
+The barrier latency is then proportional to the time it takes to process stream chunks between barriers.
+
+This is basically the number of chunks in a channel, multiplied by the time taken to process the join for it, or the UDF call latency for it.
+
+## What is Rate Limit
+
+Rate Limit simply limits the rate at which records are processed.
+It will only rate limit snapshot read / source read side of an actor.
+This means that barriers are not rate limited.
+
+## Rate Limit to keep barrier latency low
+
+### Backfill
+
+In a newly created MV, it will undergo Backfilling, to fill the MV with historical data and fresh data concurrently.
```

Comment:
After creating new jobs we'll naturally have more channels to fill in order to make back-pressure take effects. This might also be a reason why we get temporary increase in barrier latency.